### PR TITLE
port minibuffer-window and frame-root-window

### PR DESCRIPTION
**Pull request by shaleh**

### CODE REVIEW

**Comment by shaleh** [comment]
@brotzeit this fails make check and I have not had time to debug it. But this shows my refactors.

**Comment by shaleh** [comment]
My failure was `map_or` ran `frame_live_or_selected` for the 'or' case even when it was not needed. I should have used `map_or_else`. But I decided the if/else was easier to read and debug for the long term.

The code now passes make check locally.

File: rust_src/src/frames.rs [hunk 1]
```diff
@@ -1,14 +1,57 @@
...
 pub type LispFrameRef = ExternalPtr<Lisp_Frame>;
 
+impl LispFrameRef {
+    #[inline]
+    pub fn is_live(&self) -> bool {
+        unsafe { fget_terminal(self.as_ptr()) != ptr::null() }
+    }
+}
+
+pub fn frame_or_selected(object: LispObject) -> LispFrameRef {
```
Comment (birkenfeld) [inline rust_src/src/frames.rs]: Add some docstrings?

File: rust_src/src/frames.rs [hunk 2]
```diff
@@ -1,14 +1,57 @@
...
 
 use lisp::{ExternalPtr, LispObject};
 use lisp::defsubr;
 
 pub type LispFrameRef = ExternalPtr<Lisp_Frame>;
 
+impl LispFrameRef {
+    #[inline]
+    pub fn is_live(&self) -> bool {
+        unsafe { fget_terminal(self.as_ptr()) != ptr::null() }
```
Comment (birkenfeld) [inline rust_src/src/frames.rs]: should be using `ptr.is_null()` method instead?

Comment (shaleh) [reply rust_src/src/frames.rs]: Good question. I will have to try it and see what happens.

You are suggesting:

    ! fget_terminal(self.as_ptr()).is_null()

correct?

File: rust_src/src/frames.rs (hunk 1 above)
Comment (shaleh) [reply rust_src/src/frames.rs]: What could I say that is not obvious from the code? if nil return selected frame else frame

File: rust_src/src/frames.rs (hunk 2 above)
Comment (birkenfeld) [reply rust_src/src/frames.rs]: yep

**Pull request merged**